test/data_base.cpp

```diff
@@ -1,401 +1,1148 @@
+"1d05e0ed6d01010000000100000000000000000000000000000000000000000000000000000" \
+"00000000000ffffffff0704ffff001d010effffffff0100f2052a0100000043410494b9d3e7" \
+"6c5b1629ecf97fff95d7a4bbdac87cc26099ada28066c6ff1eb9191223cd897194a08d0c272" \
+"6c5747f1db49e8cf90e75dc3e3550ae9b30086f3cd5aaac00000000"
+
+class data_base_accessor
+ : public data_base
+{
+public:
+    data_base_accessor(const bc::database::settings& settings)
+        : data_base(settings)
+    {
+    }
+
+    bool push_all(block_const_ptr_list_const_ptr blocks, const config::checkpoint& fork_point)
+    {
+        return data_base::push_all(blocks, fork_point);
+    }
+
+    bool push_all(header_const_ptr_list_const_ptr headers, const config::checkpoint& fork_point)
+    {
+        return data_base::push_all(headers, fork_point);
+    }
+    
+    code push_header(const header& header, size_t height, uint32_t mtp)
+    {
+        return data_base::push_header(header, height, mtp);
+    }
+
+    code push_block(const block& block, size_t height)
+    {
+        return data_base::push_block(block, height);
+    }
+    
+    code store(const transaction& tx, uint32_t forks)
+    {
+        return data_base::store(tx, forks);
+    }
+    
+    code pop_header(chain::header& out_header, size_t height) 
+    {
+        return data_base::pop_header(out_header, height);
+    }
+
+    code pop_block(chain::block& out_block, size_t height)
+    {
+        return data_base::pop_block(out_block, height);
+    }
+
+    bool pop_above(header_const_ptr_list_ptr headers,
+                   const config::checkpoint& fork_point)
+    {
+        return data_base::pop_above(headers, fork_point);
+    }
+
+    bool pop_above(block_const_ptr_list_ptr blocks,
+        const config::checkpoint& fork_point)
+    {
+        return data_base::pop_above(blocks, fork_point);
+    }
+
+};
+
+static void
+test_heights(const data_base& instance, size_t candidate_height_in, size_t confirmed_height_in)
+{    
+    size_t candidate_height = 0u;
+    size_t confirmed_height = 0u;
+    BOOST_REQUIRE(instance.blocks().top(candidate_height, true));
+    BOOST_REQUIRE(instance.blocks().top(confirmed_height, false));
+    
+    BOOST_REQUIRE_EQUAL(candidate_height, candidate_height_in);
+    BOOST_REQUIRE_EQUAL(confirmed_height, confirmed_height_in);   
+}
+
+BOOST_AUTO_TEST_CASE(data_base__create__block_transactions_index_interaction__success)
+{
+   create_directory(DIRECTORY);
+   database::settings settings;
+   settings.directory = DIRECTORY;
+   settings.index_addresses = false;
+   settings.flush_writes = false;
+   settings.file_growth_rate = 42;
+   settings.block_table_buckets = 42;
+   settings.transaction_table_buckets = 42;
+   settings.address_table_buckets = 42;
+  
+   data_base instance(settings); 
+   
+   static const auto bc_settings = bc::settings(bc::config::settings::mainnet);
+   const chain::block genesis = bc_settings.genesis_block;
```

Comment:
`auto&` doesn't work because `bc::settings#genesis_block` returns a `config::block`, so we need to cast it to `chain::block`.

Changed all such occurrences to to `const chain::block& genesis = bc_settings.genesis_block;`. This will avoid the extraneous call to the constructor for `chain::block`